### PR TITLE
8330523: Reduce runtime and improve efficiency of KeepAliveTest

### PR DESCRIPTION
**Pull request by RealCLanger**
The test case sun/net/www/http/HttpClient/KeepAliveTest.java could be more effective.

It tests a matrix of HTTP client settings and server behavior, resulting in 160 individual test scenarios. Each is tested in an own freshly spawned JVM via the `@run main/othervm` directive. The need for new VMs is due to the fact that the behavior of the HTTP client is determined at VM initialization and can not be changed later on. However, for each distinct type of client settings, one VM can be reused. This would lead us from 160 JVM instantiations down to 16 which has a factor 10 influence on test runtime.

E.g. on my developer laptop runtime went down from ~100s to ~10s.

I also made additional cleanups/refactoring in the test.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8330523](https://bugs.openjdk.org/browse/JDK-8330523): Reduce runtime and improve efficiency of KeepAliveTest (**Bug** - P4)

### Reviewers
 * [Daniel Jeliński](https://openjdk.org/census#djelinski) (@djelinski - **Reviewer**) ⚠️ Review applies to [6dc20dc5](https://git.openjdk.org/jdk/pull/18817/files/6dc20dc5bc151207751b680339b07fa9c14b9a36)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18817/head:pull/18817` \
`$ git checkout pull/18817`

Update a local copy of the PR: \
`$ git checkout pull/18817` \
`$ git pull https://git.openjdk.org/jdk.git pull/18817/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18817`

View PR using the GUI difftool: \
`$ git pr show -t 18817`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18817.diff">https://git.openjdk.org/jdk/pull/18817.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18817#issuecomment-2061469227)

### CODE REVIEW

**Comment by dfuch** [comment]
Hi Christoph - thanks for this fix. It looks like a very nice improvement. Please give us some time to test the changes through our CI.

**Comment by djelinski** [comment]
With 800 changed lines in a now-976 file, it's practically a new test :)

Would it make sense to move the system properties to the jtreg run lines? Like:
```
@run main/othervm KeepAliveTest -Dhttp.keepAlive.time.server=100 c1
```

**Comment by RealCLanger** [comment]
> With 800 changed lines in a now-976 file, it's practically a new test :)

Yes, in fact it is. 😄 I refrained from touching the comment table that describes all the possible combinations which make 50% of the file...

> Would it make sense to move the system properties to the jtreg run lines? Like:
> 
> ```
> @run main/othervm KeepAliveTest -Dhttp.keepAlive.time.server=100 c1
> ```

I thought about that, too. I can go that route if you like - the diff will then be even larger and I should also modify the comment table...

Shall I?

**Comment by djelinski** [comment]
If that makes the end result smaller / easier to read, go for it!

**Comment by RealCLanger** [comment]
> If that makes the end result smaller / easier to read, go for it!

Here we go!

**Review by djelinski — COMMENTED** [review]
This looks reasonable.

File: test/jdk/sun/net/www/http/HttpClient/KeepAliveTest.java [hunk 1]
```diff
@@ -1042,231 +178,171 @@ public void run() {
...
+        System.setProperty("java.net.useSystemProxies", "false");
+
+        Field keepAliveField = sun.net.www.http.HttpClient.class.getDeclaredField("kac");
+        keepAliveField.setAccessible(true);
+        keepAliveCache = (KeepAliveCache) keepAliveField.get(null);
+        System.out.println("KeepAliveCache: " + keepAliveCache);
+        keepAliveKeyClassconstructor = Class.forName("sun.net.www.http.KeepAliveKey").getDeclaredConstructors()[0];
+        keepAliveKeyClassconstructor.setAccessible(true);
+
+        Logger logger = Logger.getLogger("sun.net.www.protocol.http.HttpURLConnection");
```
[the remark this answers is not in the document]
Comment (RealCLanger) [reply test/jdk/sun/net/www/http/HttpClient/KeepAliveTest.java:315]: correct. Will change it back.

File: test/jdk/sun/net/www/http/HttpClient/KeepAliveTest.java [hunk 2]
```diff
@@ -206,830 +58,114 @@
...
-     * for serverscenario[0],serverscenario[16],serverscenario[32] ... serverscenario[144] is mapped to clientscenario[0]
-     * for serverscenario[1],serverscenario[17],serverscenario[33] ... serverscenario[145] is mapped to clientscenario[1]
-     * for serverscenario[2],serverscenario[18],serverscenario[34] ... serverscenario[146] is mapped to clientscenario[2]
-     * ...
-     * for serverscenario[15],serverscenario[31],serverscenario[47] ... serverscenario[159] is mapped to clientscenario[15]
-     */
-    private int getClientScenarioNumber(int scenarioNumber) {
-        return scenarioNumber % 16 ;
-    }
+    private CountDownLatch serverLatch = new CountDownLatch(1);
```
[the remark this answers is not in the document]
Comment (RealCLanger) [reply test/jdk/sun/net/www/http/HttpClient/KeepAliveTest.java:112]: Yes, but there's always new instance per test. One could certainly change it to use a global static semaphore...

[the remark this answers is not in the document]
Comment (djelinski) [reply test/jdk/sun/net/www/http/HttpClient/KeepAliveTest.java:112]: Ah, good point, missed that.

File: test/jdk/sun/net/www/http/HttpClient/KeepAliveTest.java [hunk 3]
```diff
@@ -206,830 +58,114 @@
...
+    private Thread server;
```
[the remark this answers is not in the document]
Comment (RealCLanger) [reply test/jdk/sun/net/www/http/HttpClient/KeepAliveTest.java]: OK, did this and using virtual threads now.

**Review by djelinski — APPROVED** [review]
LGTM. Thanks!

**Comment by RealCLanger** [comment]
> LGTM. Thanks!

Thanks for the review. Will you give it another spin in your test system? Otherwise, I'll integrate tomorrow after we have another run in ours.

**Review by djelinski — APPROVED** [review]

**Comment by RealCLanger** [comment]
Testing looks good.

/integrate

**Pull request closed without merging**